utils/spaces_utils.py

```diff
@@ -0,0 +1,191 @@
+# Original work Copyright (c) 2016 OpenAI (https://openai.com).
+# Modified work Copyright (c) Allen Institute for AI
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Union, Tuple, List
+from collections import OrderedDict
+
+import numpy as np
+import torch
+from gym import spaces as gym
+
+from core.algorithms.onpolicy_sync.policy import ActionType
+
+
+def flatdim(space):
+    """Return the number of dimensions a flattened equivalent of this space
+    would have.
+
+    Accepts a space and returns an integer. Raises ``NotImplementedError`` if
+    the space is not defined in ``gym.spaces``.
+    """
+    if isinstance(space, gym.Box):
+        return int(np.prod(space.shape))
+    elif isinstance(space, gym.Discrete):
+        return 1  # we do not expand to one-hot
+    elif isinstance(space, gym.Tuple):
+        return int(sum([flatdim(s) for s in space.spaces]))
+    elif isinstance(space, gym.Dict):
+        return int(sum([flatdim(s) for s in space.spaces.values()]))
+    elif isinstance(space, gym.MultiBinary):
+        return int(space.n)
+    elif isinstance(space, gym.MultiDiscrete):
+        return int(np.prod(space.shape))
+    else:
+        raise NotImplementedError
+
+
+def flatten(space, torch_x):
+    """Flatten data points from a space.
+    """
+    if isinstance(space, gym.Box):
+        if len(space.shape) > 0:
+            return torch_x.view(torch_x.shape[: -len(space.shape)] + (-1,))
+        else:
+            return torch_x.view(torch_x.shape + (-1,))
+    elif isinstance(space, gym.Discrete):
+        # Assume tensor input does NOT contain a dimension for action
+        if isinstance(torch_x, torch.Tensor):
+            return torch_x.unsqueeze(-1)
+        else:
+            return torch.tensor(torch_x).view(1)
+    elif isinstance(space, gym.Tuple):
+        return torch.cat(
+            [flatten(s, x_part) for x_part, s in zip(torch_x, space.spaces)], dim=-1
+        )
+    elif isinstance(space, gym.Dict):
+        return torch.cat(
+            [flatten(s, torch_x[key]) for key, s in space.spaces.items()], dim=-1
+        )
```

Comment:
Are gym dictionaries automatically ordered? Just want to make sure the order we're flattening is the same every time.

Reply:
They're OrderedDicts.

Reply:
:)